Use user defined seek times, relative to current position, with optional confirmation

itayavra opened the pull request:
**Changes**

- Seeking now uses configurable forward/backward durations (5-30s) and operates relative to current position instead of snapping to predefined positions. 
- Add optional seek confirmation mode that previews position with thumbnails before applying changes. 
- Implement custom thumbnail preview system to support the new seeking behavior since it no longer uses the native player seek mechanism. 
- Add settings for backward seek time to complement the forward seek time, and seek confirmation toggle in playback preferences.

**Issues**
https://github.com/jellyfin/jellyfin-androidtv/issues/460
https://github.com/jellyfin/jellyfin-androidtv/issues/2529
https://github.com/jellyfin/jellyfin-androidtv/issues/2592
https://github.com/jellyfin/jellyfin-androidtv/issues/3276
https://github.com/jellyfin/jellyfin-androidtv/issues/3154
https://github.com/jellyfin/jellyfin-androidtv/issues/2399

itayavra committed:
Add seek functionality for playback controls in CustomPlaybackOverlayFragment

- Implemented skipping forward and backward using DPAD keys based on user preferences.

itayavra committed:
Add seek confirmation feature for D-pad navigation in playback controls

- Introduced a new user preference to require confirmation before seeking with D-pad.
- Updated CustomPlaybackOverlayFragment to handle seek preview and confirmation logic.
- Added UI elements in PlaybackPreferencesScreen for the new preference.
- Updated strings.xml with titles and descriptions for the new preference.

itayavra committed:
Remove seek confirmation feature from playback controls

- Deleted the user preference for requiring confirmation before seeking with D-pad.
- Cleaned up related logic in CustomPlaybackOverlayFragment to eliminate seek preview handling.
- Removed UI elements and bindings in PlaybackPreferencesScreen associated with the seek confirmation preference.
- Updated strings.xml to remove references to the seek confirmation feature.

itayavra committed:
Add accelerated seeking functionality to playback controls

- Introduced state management for accelerated seeking in CustomPlaybackOverlayFragment.
- Implemented methods to reset seeking state and calculate accelerated seek amounts based on consecutive seeks.
- Updated key event handling to reset seeking state when fast forward, rewind, or play/pause actions are performed.
- Enhanced skip functionality to utilize accelerated seeking logic for both forward and backward skips.

itayavra committed:
Refactor seeking functionality in CustomPlaybackOverlayFragment

- Removed accelerated seeking state management and related methods.
- Updated key event handling to directly call fast forward and rewind methods for consistent visual feedback.
- Eliminated unnecessary comments and code related to seeking state resets.
- Streamlined skip functionality by removing direct seek mode logic.

itayavra committed:
Refactor playback control methods in CustomPlaybackOverlayFragment

- Updated fast forward and rewind functionality to use VideoPlayerAdapter for consistent behavior across controls.
- Added getPlayerAdapter method in LeanbackOverlayFragment to facilitate this change.
- Improved code readability by removing outdated comments and ensuring consistent method calls.

itayavra committed:
Refactor getSafeSeekPosition method in Utils.kt

- Updated the getSafeSeekPosition method to use a minimum safe position of 100ms to prevent ExoPlayer issues with position 0.
- Simplified the logic by using coerceIn for better readability and maintainability.

itayavra committed:
Add settings to enable the enhanced D-pad seeking logic

- Introduced a new user preference for enhanced D-pad seeking in UserPreferences.kt.
- Updated CustomPlaybackOverlayFragment to utilize the new preference for improved D-pad seeking logic.
- Added UI elements in PlaybackPreferencesScreen for the enhanced D-pad seeking option.
- Updated strings.xml with titles and descriptions for the new preference.

itayavra committed:
fix: Revert unrelated changes and preserve original D-pad behavior

- Revert FF/RW key changes to playbackController methods
- Restore setInjectedViewsVisibility() when feature flag is disabled
- Update setting name to "Natural D-pad controls"

itayavra committed:
refactor: Rename to natural D-pad seeking and extract preference variable

- Update preference name and storage key to use "natural" terminology
- Extract preference check to variable for better readability
- Update comments to match new naming convention

itayavra committed:
refactor: Update comments for seeking behavior in CustomPlaybackOverlayFragment

- Revised comments for fast forward and rewind methods to clarify the purpose of using VideoPlayerAdapter.
- Removed unnecessary blank lines for improved code readability.

itayavra committed:
fix: fixed compile

itayavra committed:
refactor: code cleanups and better UX by lowering the seek debounce time

itayavra committed:
refactor: code review fixes

- Reverted changes in debounce time for seeking actions.
- Expose isSeekBarFocused method in CustomPlaybackTransportControlGlue to check seek bar focus state directly instead of checking it via reflection.

itayavra committed:
refactor: update playback control to use ProgressBar and fix lint errors

- Replaced SeekBar with ProgressBar in CustomPlaybackTransportControlGlue for better UI representation.
- Introduced constants for minimum safe seek position and end position buffer in Utils.kt to enhance seek position logic.

itayavra committed:
fix: fixed runtime error when playing video

itayavra committed:
feat: implement preview seeking functionality in playback controls

- Added support for preview seeking in CustomPlaybackOverlayFragment, allowing users to preview seek positions without affecting playback.
- Introduced methods to enter and exit preview seeking mode, and to update the preview position in PlaybackController.
- Updated UserPreferences to enable natural D-pad seeking by default.

itayavra committed:
refactor: merged the natural d-pad seeking preference as the only one and set a new preference instead to allow the user to choose if the skip will happen immediately or after confirmation

- Renamed natural D-pad seeking preference to preview seeking in UserPreferences.
- Updated CustomPlaybackOverlayFragment to support preview seeking, allowing users to preview positions before confirming.
- Adjusted UI elements in PlaybackPreferencesScreen and strings.xml to reflect the new preview seeking feature.

itayavra committed:
fix: update seeking methods to use PlayerAdapter again for improved UX

itayavra committed:
feat: add skip back length preference to playback settings

itayavra committed:
chore: removed empty lines and added nullability annotation in LeanbackOverlayFragment

itayavra committed:
fix: bugfix and preview-seek key handling improvements

- MEDIA_PLAY and MEDIA_PLAY_PAUSE now confirm the preview position and resume
  playback (same behaviour as OK / Enter).
- MEDIA_PAUSE is ignored during preview-seek; only Back / B / Escape cancel
  and return to the original position.
- Live-TV, guide, chapter overlays and trick-play should remain unaffected by this feature.
- Fixes progress bar and progress-reporting getting stuck when exiting
  preview with Play / Play-Pause.

itayavra committed:
refactor: clean up comments

itayavra committed:
refactor:  clean up preview seeking code

itayavra committed:
refactor:  clean up preview seeking code

itayavra committed:
refactor:  clean up preview seeking code

itayavra committed:
feat: added a simple thumbnail preview for trickplay when preview seeking

itayavra committed:
refactor: moved thumbnails implementation into CustomPlaybackTransportControlGlue for now

itayavra committed:
refactor: extracted code to the glue and to a new handler for thumbnail specific logic

itayavra committed:
refactor: fixed seeking on seekbar focus, simplified the glue code a bit

itayavra committed:
refactor: update thumbnail position and dimensions, enabled thumbnails also when the preview mode is turned off when trickplay is enabled

itayavra committed:
refactor: adjust thumbnail dimensions and background to hide black bars when media aspect ratio differs

itayavra committed:
refactor: code cleanup

itayavra committed:
refactor: code cleanup

itayavra committed:
refactor: code cleanup

itayavra committed:
refactor: code cleanup

itayavra committed:
refactor: code cleanup

itayavra committed:
refactor: update preview seeking confirmation text for clarity

itayavra committed:
refactor: rename preview seeking to seek confirmation for better terminology and clarity

itayavra committed:
refactor: make sure the thumbnail preview is removed if exists when playback overlay is closing with cancel button

itayavra committed:
refactor: shorter seek confirmation text

itayavra committed:
refactor: remove extra newline

itayavra commented:
@nielsvanvelzen just clarifying this is ready for review on my part

nielsvanvelzen commented on app/src/main/java/org/jellyfin/androidtv/preference/UserPreferences.kt:
```diff
@@ -252,6 +252,12 @@ class UserPreferences(context: Context) : SharedPreferenceStore(
+		 * Require confirmation before seeking to allows users to preview seek positions before confirming them.
+		 * When disabled, D-pad left/right will seek immediately.
+		 */
+		var seekConfirmationRequired = booleanPreference("seek_confirmation", true)
```
We should choose one behavior for the seek control and stick with that. No preferences to complicate it necessary.

itayavra replied:
To me, the intuitive behavior would be to pause the video when seeking, especially when thumbnail previews can be utilized, this is what I'm used to from netflix, youtube (well, smart tube), etc.
However, I saw that there were some issues open specifically about that by users that prefer the other behavior (https://github.com/jellyfin/jellyfin-androidtv/issues/2399, https://github.com/jellyfin/jellyfin-androidtv/issues/3154, ...). 
I can change that, but I would consider giving the option to choose, since it seems like it would allow more users to feel at home/enjoy what they're expecting (especially given supporting both behaviors in the code was pretty easy).
Anyway, let me know what you decide.

nielsvanvelzen commented on app/src/main/java/org/jellyfin/androidtv/ui/playback/overlay/ThumbnailPreviewHandler.java:
```diff
@@ -0,0 +1,75 @@
+
+import org.jellyfin.androidtv.R;
+
+public class ThumbnailPreviewHandler {
```
New classes must be written in Kotlin

itayavra replied:
Done

nielsvanvelzen commented on app/src/main/res/layout/vlc_player_interface.xml:
```diff
@@ -116,6 +116,18 @@
             app:layout_constraintTop_toTopOf="parent" />
     </androidx.constraintlayout.widget.ConstraintLayout>
 
+    <ImageView
```
New UI must be written in compose unless there is a good reason not to.

nielsvanvelzen commented on app/src/main/java/org/jellyfin/androidtv/ui/playback/overlay/ThumbnailPreviewHandler.java:
```diff
@@ -0,0 +1,75 @@
+        this.skipForwardLength = skipForwardLength;
+    }
+
+    public void updateThumbnailPreview(long previewSeekPosition) {
```
This doesn't properly cancel already loading images causing race conditions.

itayavra replied:
@nielsvanvelzen I thought about this issue and figured there are 2 possible race conditions here:
1. Multiple requests for thumbnails/tiles from the same sheet (e.g. tiles 0-99 all fetching the same tile sheet simultaneously)
2. Callbacks returning in wrong order, showing outdated thumbnails during fast scrubbing

For the first issue - I think the `ImageLoader` in `CustomSeekProvider` handles caching/deduplication network requests? Although even if not then the tile sheets are relatively small, so this shouldn't be too heavy on the network. 
As an optimization I added a prefetch for the sheet that includes the tiles near the starting position of the media.

For the second issue (which I assume you were referring to) - I tested simply canceling all previous requests, but this creates terrible UX during fast scrubbing since very few thumbnails actually get shown while the user is moving.

In the end I chose the following, which improved performance a lot in my tests:
- Ignore callbacks from older requests (meaning, we already showed a thumbnail from a newer request)
- Ignore stale requests, that are 20+ requests behind (this threshold gave the best feel in my opinion)

nielsvanvelzen reviewed (changes requested):
Initial review based on a quick look over the code and testing in emulator. I'm currently unsure if this is the correct approach or if we should replace the entire seekbar UI to avoid hacking into leanback code.

itayavra commented:
> Initial review based on a quick look over the code and testing in emulator. I'm currently unsure if this is the correct approach or if we should replace the entire seekbar UI to avoid hacking into leanback code.

Not sure what the implication/complexity of that would be since I'm too new here, but if this is not something that is planned anytime soon then perhaps one does not negate the other?

itayavra committed:
cr fixes - converted ThumbnailPreviewHandler to kt
more precide intervals for the seek positions/thumbnails
replaced thumbnail element to Compose

itayavra committed:
refactor:
- added prefetching of relevant tile sheet in case trickplay is enabled to reduce chance of multiple HTTP requests for tiles in the same sheet around the playback position on play
- fixed thumbnail compose look
- performance improvements in ThumbnailPreviewHandler:
-- ignoring a thumbnail request if it's too far away from the current request (and just returned with a big delay)
-- ignoring a thumbnail request if we already displayed a thumbnail that came from a newer request

itayavra commented:
@nielsvanvelzen how would you like to proceed with this?

lgruen commented on app/src/main/java/org/jellyfin/androidtv/ui/playback/overlay/ThumbnailPreviewHandler.kt:
```diff
@@ -0,0 +1,70 @@
+                if (bitmap == null) return
+                
+                // Ignore requests that are older than what we last showed
+                if (currentRequestNumber <= lastShownRequestNumber.get()) return
```
This guard is the seek-preview "black box" bug — the real thumbnail never replaces the placeholder.

`CustomSeekProvider.getThumbnail()` fires the `ResultCallback` **twice** for a single request: first from coil's `onStart` with the placeholder bitmap (the `black_transparent_light` one added in #4946), then from `onSuccess` with the real frame — both carrying the same `currentRequestNumber`.

The `onStart` placeholder arrives first, passes this guard, and sets `lastShownRequestNumber = currentRequestNumber` below. When `onSuccess` then delivers the real frame for the **same** request, `currentRequestNumber <= lastShownRequestNumber` is now true, so it returns early and the real thumbnail is discarded — leaving a permanent translucent-black box (deterministic on any cache-miss, since `onStart` always precedes `onSuccess`).

The comment above says *"older than what we last showed"*, but `<=` also rejects the *equal* (same-request) `onSuccess`. Strict `<` matches that intent — a request can still replace its own placeholder, while genuinely older requests are still dropped during fast scrubbing:

```suggestion
                if (currentRequestNumber < lastShownRequestNumber.get()) return
```

(The native leanback scrub path is unaffected, since it uses leanback's own `ResultCallback`, which renders whatever bitmap it receives.) Confirmed on an Nvidia Shield build cherry-picking this PR onto current `master` — this one-line change restores the previews.
